apps/s_socket.c

```diff
@@ -322,9 +325,17 @@ int do_server(int *accept_sock, const char *host, const char *port,
                 ERR_print_errors(bio_err);
                 goto end;
             }
+#if defined(__APPLE__) && defined(__MACH__)
+            /* Work around a Mac OS X quirk... */
+            {
+                struct pollfd fd = { asock, POLLIN, 0 };
+
+                (void)poll(&fd, 1, -1);
+            }
+#endif
             do {
                 sock = BIO_accept_ex(asock, ourpeer, 0);
-            } while (sock < 0 && BIO_sock_should_retry(sock));
+            } while (sock < 0 && BIO_sock_should_retry(asock));
```

Comment:
should_retry(asock) is wrong.

Reply:
Hmmm... Right you are! Originally it stung the eye because there is nothing you can say about negative file descriptor. But apparently BIO_sock_should_retry uses it only to determine if get_last_socket_error is to be examined, so that sock < 0 is the only meaningful input. Sorry about confusion, I will revoke the commit...

BTW, on related note one can wonder if BIO_sock_should_retry is right about checking for input descriptor being 0. This effectively means that stdin being used as socket might result in some odd behaviour. stdin as socket is not actually uncommon if binary is started by inetd, is it?

Reply:
Yes, 0 is a valid socket.
A deja-vu is I fixed the very same line already before daaaa3cb :-)